model/task_lifecycle.go

```diff
@@ -992,36 +992,70 @@ func RestartFailedTasks(opts RestartOptions) (RestartResults, error) {
 	if opts.IncludeSetupFailed {
 		failureTypes = append(failureTypes, evergreen.CommandTypeSetup)
 	}
-	tasksToRestart, err := task.Find(task.ByTimeStartedAndFailed(opts.StartTime, opts.EndTime, failureTypes))
+	tasksToRestart, err := task.FindWithDisplayTasks(task.ByTimeStartedAndFailed(opts.StartTime, opts.EndTime, failureTypes))
 	if err != nil {
 		return results, err
 	}
 
+	type taskGroupAndBuild struct {
+		Build     string
+		TaskGroup string
+	}
+	// only need to check one task per task group / build combination, and once per display task
+	taskGroupsToCheck := map[taskGroupAndBuild]string{}
+	displayTasksToCheck := map[string]*task.Task{}
+	idsToRestart := []string{}
+	for _, t := range tasksToRestart {
+		if t.IsPartOfDisplay() {
+			displayTasksToCheck[t.DisplayTask.Id] = t.DisplayTask
+		} else if t.DisplayOnly {
+			displayTasksToCheck[t.Id] = t.DisplayTask
+		} else if t.IsPartOfSingleHostTaskGroup() {
+			taskGroupsToCheck[taskGroupAndBuild{
+				TaskGroup: t.TaskGroup,
+				Build:     t.BuildId,
+			}] = t.Id
+		} else {
+			idsToRestart = append(idsToRestart, t.Id)
+		}
+	}
+
+	for id, dt := range displayTasksToCheck {
+		if dt.IsFinished() {
```

Comment:
This seems different from the logic in [checkResetDisplayTask](https://github.com/evergreen-ci/evergreen/blob/d18f80b679a965bac7d994751c25e2d7fbb29900/model/task_lifecycle.go#L1217)

1. there's an exception there for a deactivated execution task
2. there we determine a display task is finished by checking each of its constituent execution tasks (though this is probably equivalent to sorting by status priority for the case of a finished display task)

Reply:
This logic is different, yes. But a deactivated execution task will still mark the display task as finished, right? Or if not then maybe we don't care to restart that display task, since this only gets hit when an admin restarts a group of tasks from the admin page. So this does still seem safe.